autotelic/ietf-hmac-signatures
# Feat: support typescript

### Summary
resolves #18 
resolves https://github.com/telus/sre-risk-engine/issues/658

- Installed typescript as a dev dependency so we can automatically define typescript types based on jsdoc comments
  - This allows the type definitions to be co-located with the actual JS code and therefore may be more likely to be updated in conjunction with the code
- Set up tools with new typescript support
- Added JSdoc definitions to `src/index.js` 
  - running `npm run tsc` or `npm test` will now automatically generate typescript definitions from these comments. 
- Note: all files in `/types` are automatically generated by `tsc` with the exception of `types/index.test-d.ts`  
- Created basic typescript tests in `types/index.test-d.ts` for `tsd` to run part of the `npm test` command

### References 
- https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html
- https://classic.yarnpkg.com/en/docs/cli/link/

### Test Plan
- Clone, checkout, and install
- Run local tests `npm run tsc:test`
- Confirm there are no `tsc `or `tsd` errors
- Set up yarn link in the root of `ietf-hmac-signatures` 
  - run `yarn link`  
- Clone & checkout the `feat/import-ietf-types` branch from https://github.com/autotelic/developer-portal/pull/7
- Set up and install local developer portal instance `yarn install`
- Reproduce typescript error `yarn tsc`
- link the `ietf-hmac-signatures` dependency to the local checked out instance
  - run `yarn link @autotelic/ietf-hmac-signatures` in the root of the `developer-portal` application
- Confirm new types remove typescript errors `yarn tsc`
- Run the test plan in https://github.com/autotelic/developer-portal/pull/7 to confirm the application runs properly with the new type definitions.

Clean-Up:
- In the root of the `developer-portal` application, run `yarn unlink @autotelic/ietf-hmac-signatures && yarn install --force`

### TODO (post-approval)
- Merge changes and create new next release
- Update the changes in https://github.com/autotelic/developer-portal/pull/7 to use the newly released version
